books/RayTracingInOneWeekend.html

```diff
@@ -2779,6 +2782,239 @@
+                ray scattered;
+                color attenuation;
+                if (rec.mat_ptr->scatter(r, rec, attenuation, scattered))
+                    return attenuation * ray_color(scattered, depth-1);
+                return color(0,0,0);
+            }
+
+            vec3 unit_direction = unit_vector(r.direction());
+            auto t = 0.5*(unit_direction.y() + 1.0);
+            return (1.0-t)*color(1.0, 1.0, 1.0) + t*color(0.5, 0.7, 1.0);
+        }
+    };
+
+
+    #endif
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    [Listing [scene-initial]: <kbd>[scene.h]</kbd> The new scene class]
+
+</div>
+
+</div class='together'>
+With the new scene class, we trim a lot of code from main.cc. This is what's left:
+
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++
+    #include "rtweekend.h"
+
+    #include "camera.h"
+    #include "color.h"
+    #include "hittable_list.h"
+    #include "material.h"
+    #include "scene.h"
+    #include "sphere.h"
+
+    void main() {
+        scene scene_desc;
+
+        const double aspect_ratio = 16.0 / 9.0;
+
+        scene_desc.image_width = 400;
+        scene_desc.image_height = static_cast<int>(scene_desc.image_width / aspect_ratio);
+        scene_desc.samples_per_pixel = 100;
+
+        hittable_list& world = scene_desc.world;
+
+        auto material_ground = make_shared<lambertian>(color(0.8, 0.8, 0.0));
+        auto material_center = make_shared<lambertian>(color(0.1, 0.2, 0.5));
+        auto material_left   = make_shared<dielectric>(1.5);
+        auto material_right  = make_shared<metal>(color(0.8, 0.6, 0.2), 0.0);
+
+        world.add(make_shared<sphere>(point3( 0.0, -100.5, -1.0), 100.0, material_ground));
+        world.add(make_shared<sphere>(point3( 0.0,    0.0, -1.0),   0.5, material_center));
+        world.add(make_shared<sphere>(point3(-1.0,    0.0, -1.0),   0.5, material_left));
+        world.add(make_shared<sphere>(point3(-1.0,    0.0, -1.0),  -0.4, material_left));
+        world.add(make_shared<sphere>(point3( 1.0,    0.0, -1.0),   0.5, material_right));
+
+        scene_desc.render();
+    }
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    [Listing [main-with-scene]: <kdb>[main.cc]</kbd> main.cc using the new scene class]
+
+</div>
+
+Run the new program, you should get the same result as image 16 again.
+
+
+Refactoring Camera Setup
+-------------------------
+Our camera has values that are either hard-coded, or can only be specified on construction. That
+makes things difficult with the scene class, particularly if we want to start making scene-creating
+functions that also set up the camera. To make this easier, we'll allow camera settings to be
+altered at will, and then perform the camera initialization explicitly with a call to
+`camera::initialize()`.
```

Comment:
We should stop and explain why initialization is very important.

Since we're decoupling the camera init from it's constructor, there is some private member variables that are determined once the programmer explicitly says that the public variables have been set.

If a reader doesn't fully understand what the `camera::initizalize()` function does, they could mistake it for a constructor of sorts, and do something like:

```
cam.intialize();
cam.fov = 40;
```

Reply:
Good point.

Reply:
I added an additional sentence, but the above is very unlikely, since we have an explicit `render()` function, which initializes the camera prior to sampling the scene.